### PR TITLE
Pixelcnn

### PR DESCRIPTION
**Pull request by skeydan**

### CODE REVIEW

File: _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+The basic idea in PixelCNN is autoregressivity. Each pixel is modeled as depending on all prior pixels. Formally:
+
+$$p(\mathbf{x}) = \prod_{i}p(x_i|x_0, x_1, ..., x_{i-1})$$
+
+Now wait a second - what even *are* prior pixels? Last I saw one images were two-dimensional. So this means we have to impose
+an *order* on the pixels. Commonly this will be *raster scan* order: row after row, from left to right. But when dealing with
+color images, there's something else: At each position, we actually have *three* intensity values, one for each of red, green,
+and blue. The original PixelCNN paper[@OordKK16] carried through autoregressivity here as well, with a pixel's intensity for
+red depending on just prior pixels, those for green depending on these same prior pixels but additionally, the current value
+for red, and those for blue depending on the prior pixels as well as the current values for red and green.
```
Comment (batpigandme) [inline _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd:61]: Ah, so the "addition" is of another sort of "prior" value? I think you might be explaining this as well as is humanly possible, I just know very very little about this, and autoregressivity is hard to describe! 😆

File: _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+  generated_images <- generated_images %>% as.array()
+  generated_images <- generated_images[ , , , 1]/255
+  
+  images <- abind::abind(real_images, generated_images, along = 1)
+  png(paste0("draw_", i, ".png"), width = 8 * 28 * 10, height = 2 * 28 * 10)
+  par(mfrow = c(2, 8), mar = c(0, 0, 0, 0))
+  images %>%
+    purrr::array_tree(1) %>%
+    purrr::map(as.raster) %>%
+    purrr::iwalk(~{plot(.x)})
```
Comment (javierluraschi) [inline _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd:295]: Can't we remove `{}`? As in `purrr::iwalk(~plot(.x))`, or even better, `purrr::iwalk(plot)`?

File: _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+
+Technically, then, we know how autoregressivity is realized; intuitively, it may still seem surprising that imposing a raster
+scan order "just works" (to me, at least, it is). Maybe this is one of those points where compute power successfully
+compensates for lack of an equivalent of a cognitive prior.
+
+### Masking, or: Where not to look
+
+Now, PixelCNN ends in "CNN" for a reason -- as usual in image processing, convolutional layers (or blocks thereof) are
+involved. But -- is it not the very nature of a convolution that it computes an average of some sorts, looking, for each
+position, not just at its own value but also, at its spatial (or temporal) surroundings? How does that rhyme with the
```
Comment (batpigandme) [inline _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd:78]:
```suggestion
position not just at its own value but also, at its spatial (or temporal) surroundings? How does that rhyme with the
```

File: _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+[^2]: For details, see [@Salimans2017PixeCNN].
+
+Technically, then, we know how autoregressivity is realized; intuitively, it may still seem surprising that imposing a raster
+scan order "just works" (to me, at least, it is). Maybe this is one of those points where compute power successfully
+compensates for lack of an equivalent of a cognitive prior.
+
+### Masking, or: Where not to look
+
+Now, PixelCNN ends in "CNN" for a reason -- as usual in image processing, convolutional layers (or blocks thereof) are
+involved. But -- is it not the very nature of a convolution that it computes an average of some sorts, looking, for each
```
Comment (batpigandme) [inline _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd:77]:
```suggestion
involved. But -- is it not the very nature of a convolution that it computes an average of some sorts, looking for each
```

File: _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+`num_resnet` and `num_hierarchies` are the parameters you're most likely to experiment with, but there are a few more you can
+check out in the [documentation](https://rstudio.github.io/tfprobability/reference/tfd_pixel_cnn.html). The number of logistic
+distributions in the mixture is also configurable, but from my experiments it's best to keep that number rather low to avoid
+producing `NaN`s during training.
+
+Let's now see a complete example.
+
+## End-to-end example
+
+Our playground will be [QuickDraw](tbd), a dataset -- still growing -- obtained by asking people to draw some object in at
```
Comment (batpigandme) [inline _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd:162]: Just a reminder that you've got the ***tbd*** for the link here!

File: _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+![Fig. 6: Butterflies, drawn by people (top row) and the network (bottom row).](images/draw_49.png)
+
+![Fig. 7: Guitars, drawn by people (top row) and the network (bottom row).](images/draw_134.png)
+
+![Fig. 8: Penguins, drawn by people (top row) and the network (bottom row).](images/draw_218.png)
+
+![Fig. 9: Roller skates, drawn by people (top row) and the network (bottom row).](images/draw_246.png)
+
+We probably wouldn't confuse the first and second rows, but then, the actual human drawings exhibit enormous variation, too.
+And no-one ever said PixelCNN was an architecture for concept learning. Feel free to play around with other datasets of your
```
Comment (batpigandme) [inline _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd:315]:
```suggestion
And no one ever said PixelCNN was an architecture for concept learning. Feel free to play around with other datasets of your
```

Comment (batpigandme) [reply _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd:315]: I don't hyphenate no one. 🙂

File: _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd (hunk 2 above)
Comment (skeydan) [reply _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd]: that's a lot nicer, thanks!

File: _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+We've seen quite a few examples of unsupervised learning (or self-supervised learning, to choose the more correct but less
+popular term) on this blog.
+
+Often, these involved *Variational Autoencoders (VAEs),* whose appeal lies in them allowing to model a *latent space* of
+underlying, independent (preferably) factors that determine the visible features. A possible downside can be the inferior
+quality of generated samples. Another popular approach are GANs (Generative Adversarial Networks). Conceptually, these are
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd:33]: you're totally right :-)

File: _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+highly attractive due to their game-theoretic framing. However, they can be difficult to train. *PixelCNN* variants, on the
+other hand -- we'll subsume them all here under PixelCNN -- are generally known for their good results. They seem to involve
+some more alchemy[^1] though. Under those circumstances, what could be more welcome than an easy way of experimenting with
+them? Through TensorFlow Probability (TFP) and its R wrapper, [tfprobability](https://github.io/tfprobability), we now have
+such a way.
+
+[^1]: Alluding to Ali Rahimi's (in-)famous "deep learning is alchemy" [talk](https://www.youtube.com/watch?v=Qi1Yry33TQE) at
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd:40]: guess I wanted to have it a bit ambiguous here, famous and/or infamous, what have you...
But I'll go with infamous as you suggest, that should work as well...

File: _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd (hunk 1 above)
Comment (skeydan) [reply _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd:61]: don't know - I think it's consistent in the context...

File: _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd (hunk 3 above)
Comment (skeydan) [reply _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd:78]: not sure about this one? I think it needs the comma for the content to be correct...

File: _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd (hunk 4 above)
Comment (skeydan) [reply _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd:77]: same here... I would think the comma is needed ... perhaps the word "position" is a bad choice here (and above)... Perhaps I should call this "for each convolution", or even, "for each convolution output" ...

File: _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd (hunk 5 above)
Comment (skeydan) [reply _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd:162]: ooops ... thanks!

File: _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd (hunk 6 above)
Comment (skeydan) [reply _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd:315]: lol :-)))

File: _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd (hunk 4 above)
Comment (skeydan) [reply _posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd:77]: yeah, I think the replacement made it clearer now

**Comment by skeydan** [comment]
thanks a lot Mara, learned a lot as always :-))

**Pull request merged**